### PR TITLE
Refactor binary

### PR DESCRIPTION
**Pull request by Aloso**
This is my attempt at refactoring the binary crate. It's not finished, but I think the remaining work can be done in subsequent PRs.

### CODE REVIEW

**Comment by Aloso** [comment]
I resolved the merge conflicts by merging `main` into this branch.

@adam-mcdaniel Could you please provide feedback? I don't want to rush you, I understand it's time-consuming to review code.

File: src/binary/init_environment/rand_module.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
+use common_macros::b_tree_map;
+use dune::{Environment, Error, Expression};
+use rand::{distributions::Uniform, prelude::SliceRandom, Rng};
+
+pub fn get() -> Expression {
+    fn int(args: Vec<Expression>, env: &mut Environment) -> Result<Expression, Error> {
```
Comment (adam-mcdaniel) [inline src/binary/init_environment/rand_module.rs]: I would prefer if these functions were defined outside of `get`, since they're not closures

**Comment by adam-mcdaniel** [comment]
Also, I think it would be good to rename `init_environment` to just `init`, and then move `init_environment.rs` into `init/mod.rs`

**Comment by adam-mcdaniel** [comment]
Looks good!!! Thank you so much! :)

**Pull request merged**